hmalloc/doc/hmctl.8
.\" Automatically generated by Pandoc 2.9.2.1
.\"
.TH "HMCTL" "8" "Apr, 2024" "Hmctl User Manuals" ""
.hy
.SH NAME
.PP
hmctl - Control heterogeneous memory allocation policy
.SH SYNOPSIS
.PP
hmctl [\f[I]options\f[R]] COMMAND [\f[I]command-options\f[R]]
.SH DESCRIPTION
.PP
The \f[B]hmctl\f[R] tool is to control heterogeneous memory allocation
policy.
It changes memory policy only for the area allocated by \f[B]hmalloc
APIs\f[R], which is called as \f[B]hmalloc pool\f[R] then executes the
given \f[C]COMMAND\f[R] program.
The rest of memory regions other than \f[B]hmalloc pool\f[R] are not
affected by this tool.
.PP
\f[I]nodes\f[R] may be specified as N,N,N or N-N or N,N-N or N-N,N-N and
so forth.
Please see its annotation details at \f[B]numactl\f[R](8).
.SH OPTIONS
.TP
-m \f[I]nodes\f[R], --membind=\f[I]nodes\f[R]
Allocate memory only from \f[I]nodes\f[R] for hmalloc family
allocations.
.TP
-p \f[I]node\f[R], --preferred=\f[I]node\f[R]
Preferably allocate memory on \f[I]node\f[R] for hmalloc family
allocations.
.TP
-P \f[I]nodes\f[R], --preferred-many=\f[I]nodes\f[R]
Preferably allocate memory on \f[I]nodes\f[R] for hmalloc family
allocations.
\f[I]nodes\f[R] are a mask of preferred nodes where the closest node to
local is considered most preferred.
.TP
-i \f[I]nodes\f[R], --interleave=\f[I]nodes\f[R]
Set a memory interleave policy.
Memory will be allocated using round robin on \f[I]nodes\f[R].
.TP
-w \f[I]nodes\f[R], --weighted-interleave=\f[I]nodes\f[R]
Set a weighted memory interleave policy.
Memory will be allocated using the weighted ratio for each node, which
can be read from /sys/kernel/mm/mempolicy/weighted_interleave/node*.
.TP
-?, --help
Print help message and list of options with description
.TP
--usage
Print usage string
.SH EXAMPLES
.PP
Let\[cq]s say if the target test program allocates 512 MiB using
\f[B]hmalloc\f[R](3), then this memory area can be allocated to the
intended node as follows.
.IP
.nf
\f[C]
# Allocate hmalloc area to node 2 with MPOL_BIND policy.
$ hmctl -m 2 ./prog

# Allocate hmalloc area to node 1 with MPOL_PREFERRED policy.
$ hmctl -p 1 ./prog
\f[R]
.fi
.PP
If you want to use a different memory policy with global policy, then
hmctl can be used along with numactl as follows.
.IP
.nf
\f[C]
# Set global policy to MPOL_PREFERRED for node 1,
# but use MPOL_BIND for node 2 only to hmalloc area.
$ numactl -p 1 hmctl -m 2 ./prog

# Set global policy to MPOL_INTERLEAVE for node 1 and 2,
# but use MPOL_PREFERRED for node 3 only to hmalloc area.
$ numactl -i 1,2 hmctl -p 3 ./prog
\f[R]
.fi
.SH GLOSSARY
.SS HMALLOC APIS
.PP
The \f[B]hmalloc APIs\f[R] are heterogeneous memory allocation APIs
provided by \f[B]libhmalloc.so\f[R] such as \f[B]hmalloc\f[R](3),
\f[B]hcalloc\f[R](3), \f[B]hposix_memalign\f[R](3), \f[B]hmmap\f[R](3),
etc.
All the APIs defined in \f[B]hmalloc.h\f[R] are \f[B]hmalloc APIs\f[R].
.SS HMALLOC POOL
.PP
The \f[B]hmalloc pool\f[R] is specially managed memory areas that can be
optionally controlled by \f[B]hmctl\f[R](8) tool.
If target programs allocate memory using \f[B]hmalloc APIs\f[R], then
this area is mapped as \f[B]hmalloc pool\f[R].
This \f[B]hmalloc pool\f[R] has no effect if the target program runs
without \f[B]hmctl\f[R](8), but if it runs with \f[B]hmctl\f[R](8)
attached, then the memory policy of this area can be changed based on
the usage of \f[B]hmctl\f[R](8).
.SS HMCTL
.PP
The \f[B]hmctl\f[R](8) is a tool that controls heterogeneous memory
allocation policy.
That means it can change the memory policy of \f[B]hmalloc pool\f[R]
allocated by \f[B]hmalloc APIs\f[R] internally using \f[B]mmap\f[R](2)
and \f[B]mbind\f[R](2).
If \f[B]hmctl\f[R](8) is attached and one of memory policy option is
given with valid \f[I]node\f[R] or \f[I]nodes\f[R] argument, then the
\f[B]hmalloc pool\f[R] memory is allocated from the target node with the
given memory policy based on the usage of \f[B]hmctl\f[R](8).
.SH SEE ALSO
.PP
\f[B]numactl\f[R](8), \f[B]hmalloc\f[R](3)
.SH AUTHORS
Honggyu Kim <honggyu.kim@sk.com>.


hmalloc/src/hmctl.c
/* Copyright (c) 2024 SK hynix, Inc. */
/* SPDX-License-Identifier: BSD 2-Clause */

#include <argp.h>
#include <numa.h>
#include <numaif.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#ifndef MPOL_PREFERRED_MANY
#define MPOL_PREFERRED_MANY 5
#endif

#ifndef MPOL_WEIGHTED_INTERLEAVE
#define MPOL_WEIGHTED_INTERLEAVE 6
#endif

struct opts {
    int idx;
    char *exename;

    const char *membind;
    const char *preferred_many;
    const char *interleave;
    const char *weighted_interleave;
    int preferred;
};

struct opts opts;

/* (a part of) output in --help option (generated by argp runtime) */
const char *argp_program_bug_address = "https://github.com/skhynix/hmsdk/issues";

static struct argp_option hmctl_options[] = {
    {.name = "preferred",
     .key = 'p',
     .arg = "node",
     .doc = "Preferably allocate memory on node for hmalloc family allocations"},
    {.name = "preferred-many",
     .key = 'P',
     .arg = "nodes",
     .doc = "Preferably allocate memory on nodes for hmalloc family allocations, but allow fall "
            "back to other nodes"},
    {.name = "membind",
     .key = 'm',
     .arg = "nodes",
     .doc = "Only allocate memory from nodes for hmalloc family allocations"},
    {.name = "interleave",
     .key = 'i',
     .arg = "nodes",
     .doc = "Set a memory interleave policy. Memory will be allocated using round robin on nodes"},
    {.name = "weighted-interleave",
     .key = 'w',
     .arg = "nodes",
     .doc = "Allocate memory on nodes using the weights specified in "
            "/sys/kernel/mm/mempolicy/weighted_interleave/*"},
    {NULL},
};

static void fail_mpol_conflict(struct argp_state *state, char key) {
    fprintf(stderr, "Error: '-%c' conflicts with other memory policy.\n\n", key);
    argp_state_help(state, state->out_stream, ARGP_HELP_STD_HELP);
}

static error_t parse_option(int key, char *arg, struct argp_state *state) {
    struct opts *opts = state->input;

    switch (key) {
    case 'm':
        opts->membind = arg;
        if (opts->preferred > 0)
            fail_mpol_conflict(state, key);
        break;

    case 'P':
        opts->preferred_many = arg;
        if (opts->membind)
            fail_mpol_conflict(state, key);
        break;

    case 'p':
        opts->preferred = atoi(arg);
        if (opts->membind)
            fail_mpol_conflict(state, key);
        break;

    case 'i':
        opts->interleave = arg;
        break;

    case 'w':
        opts->weighted_interleave = arg;
        break;

    case ARGP_KEY_ARG:
        if (state->arg_num)
            return ARGP_ERR_UNKNOWN;
        if (opts->exename == NULL) {
            /* remaining options will be processed in ARGP_KEY_ARGS */
            return ARGP_ERR_UNKNOWN;
        }
        break;

    case ARGP_KEY_ARGS:
        /* process remaining non-option arguments */
        opts->exename = state->argv[state->next];
        opts->idx = state->next;
        break;

    case ARGP_KEY_NO_ARGS:
    case ARGP_KEY_END:
        if (state->arg_num < 1)
            argp_usage(state);
        break;

    default:
        return ARGP_ERR_UNKNOWN;
    }
    return 0;
}

static void setup_child_environ(struct opts *opts) {
    unsigned long nodemask = 0;
    char buf[4096] = "0";
    struct bitmask *bm;

    if (opts->membind) {
        snprintf(buf, sizeof(buf), "%d", MPOL_BIND);
        setenv("HMALLOC_MPOL_MODE", buf, 1);

        bm = numa_parse_nodestring(opts->membind);
        if (bm) {
            snprintf(buf, sizeof(buf), "%lu", *bm->maskp);
            setenv("HMALLOC_NODEMASK", buf, 1);
        }
    } else if (opts->preferred_many) {
        snprintf(buf, sizeof(buf), "%d", MPOL_PREFERRED_MANY);
        setenv("HMALLOC_MPOL_MODE", buf, 1);

        bm = numa_parse_nodestring(opts->preferred_many);
        if (bm) {
            snprintf(buf, sizeof(buf), "%lu", *bm->maskp);
            setenv("HMALLOC_NODEMASK", buf, 1);
        }
    } else if (opts->preferred >= 0) {
        /* ignore when --membind is used */
        snprintf(buf, sizeof(buf), "%d", MPOL_PREFERRED);
        setenv("HMALLOC_MPOL_MODE", buf, 1);

        nodemask = 1 << opts->preferred;
        snprintf(buf, sizeof(buf), "%ld", nodemask);
        setenv("HMALLOC_NODEMASK", buf, 1);
    } else if (opts->weighted_interleave) {
        snprintf(buf, sizeof(buf), "%d", MPOL_WEIGHTED_INTERLEAVE);
        setenv("HMALLOC_MPOL_MODE", buf, 1);

        bm = numa_parse_nodestring(opts->weighted_interleave);
        if (bm) {
            snprintf(buf, sizeof(buf), "%lu", *bm->maskp);
            setenv("HMALLOC_NODEMASK", buf, 1);
        }
    } else if (opts->interleave) {
        snprintf(buf, sizeof(buf), "%d", MPOL_INTERLEAVE);
        setenv("HMALLOC_MPOL_MODE", buf, 1);

        bm = numa_parse_nodestring(opts->interleave);
        if (bm) {
            snprintf(buf, sizeof(buf), "%lu", *bm->maskp);
            setenv("HMALLOC_NODEMASK", buf, 1);
        }
    }

    setenv("HMALLOC_JEMALLOC", "1", 1);
}

int main(int argc, char *argv[]) {
    struct argp argp = {
        .options = hmctl_options,
        .parser = parse_option,
        .args_doc = "[<program>]",
        .doc = "hmctl -- Control heterogeneous memory allocation policy",
    };

    /* default option values */
    opts.membind = NULL;
    opts.preferred = -1;

    argp_parse(&argp, argc, argv, ARGP_IN_ORDER, NULL, &opts);

    /* pass only non-hmctl options to execv() */
    argc -= opts.idx;
    argv += opts.idx;

    setup_child_environ(&opts);

    execv(opts.exename, argv);
    perror(opts.exename);

    return -1;
}
